numpy/linalg/umath_linalg.c.src

```diff
@@ -3235,160 +3269,1009 @@ static void
-    (void *)NULL,
+static inline void
+dump_geqrf_params(const char *name,
+                  GEQRF_PARAMS_t *params)
+{
+    TRACE_TXT("\n%s:\n"\
 
-    (void *)NULL,
-    (void *)NULL,
-    (void *)NULL,
-    (void *)NULL
-};
+              "%14s: %18p\n"\
+              "%14s: %18p\n"\
+              "%14s: %18p\n"\
+              "%14s: %18d\n"\
+              "%14s: %18d\n"\
+              "%14s: %18d\n"\
+              "%14s: %18d\n",
 
-#define FUNC_ARRAY_NAME(NAME) NAME ## _funcs
+              name,
 
-#define GUFUNC_FUNC_ARRAY_REAL(NAME)                    \
-    static PyUFuncGenericFunction                       \
-    FUNC_ARRAY_NAME(NAME)[] = {                         \
-        FLOAT_ ## NAME,                                 \
-        DOUBLE_ ## NAME                                 \
-    }
+              "A", params->A,
+              "TAU", params->TAU,
+              "WORK", params->WORK,
 
-#define GUFUNC_FUNC_ARRAY_REAL_COMPLEX(NAME)            \
-    static PyUFuncGenericFunction                       \
-    FUNC_ARRAY_NAME(NAME)[] = {                         \
-        FLOAT_ ## NAME,                                 \
-        DOUBLE_ ## NAME,                                \
-        CFLOAT_ ## NAME,                                \
-        CDOUBLE_ ## NAME                                \
-    }
+              "M", (int)params->M,
+              "N", (int)params->N,
+              "LDA", (int)params->LDA,
+              "LWORK", (int)params->LWORK);
+}
 
-/* There are problems with eig in complex single precision.
- * That kernel is disabled
+/**begin repeat
+   #lapack_func=sgeqrf,dgeqrf,cgeqrf,zgeqrf#
  */
-#define GUFUNC_FUNC_ARRAY_EIG(NAME)                     \
-    static PyUFuncGenericFunction                       \
-    FUNC_ARRAY_NAME(NAME)[] = {                         \
-        FLOAT_ ## NAME,                                 \
-        DOUBLE_ ## NAME,                                \
-        CDOUBLE_ ## NAME                                \
+
+static inline fortran_int
+call_@lapack_func@(GEQRF_PARAMS_t *params)
+{
+    fortran_int rv;
+    LAPACK(@lapack_func@)(&params->M, &params->N,
+                          params->A, &params->LDA,
+                          params->TAU,
+                          params->WORK, &params->LWORK,
+                          &rv);
+    return rv;
+}
+
+/**end repeat**/
+
+/**begin repeat
+   #TYPE=FLOAT,DOUBLE#
+   #lapack_func=sgeqrf,dgeqrf#
+   #ftyp=fortran_real,fortran_doublereal#
+ */
+static inline int
+init_@lapack_func@(GEQRF_PARAMS_t *params,
+                   fortran_int m,
+                   fortran_int n)
+{
+    npy_uint8 *mem_buff = NULL;
+    npy_uint8 *mem_buff2 = NULL;
+    npy_uint8 *a, *tau, *work;
+    fortran_int min_m_n = fortran_int_min(m, n);
+    size_t safe_min_m_n = min_m_n;
+    size_t safe_m = m;
+    size_t safe_n = n;
+
+    size_t a_size = safe_m * safe_n * sizeof(@ftyp@);
+    size_t tau_size = safe_min_m_n * sizeof(@ftyp@);
```

Comment:
Can these overflow if m, n are too large?

Reply:
If `a_size` will overflow then possibly the error would also be there in other functions defined in this file, since memory is allocated using similar logic everywhere. In other words, this is a broader issue and can arise in any function if a very large matrix is passed. Please let me know if I am missing out on something here. Thanks.

Reply:
For example, in the below init_lapack function for SVD, a_size is calculated in the same manner,

https://github.com/numpy/numpy/blob/21838b4b21d8c702730aba80131fe6ee5fe5116f/numpy/linalg/umath_linalg.c.src#L2479

I think if `tau_size` will overflow then `a_size` would definitely overflow.

Reply:
I am resolving this for now. Please let me know if a separate issue should be opened for this as it covers many functions in this file.